src/server/shell/abstract_shell.cpp

```diff
@@ -406,102 +432,114 @@ void msh::AbstractShell::set_focus_to(
 {
     std::unique_lock<std::mutex> lock(focus_mutex);
 
-    notify_focus_locked(lock, focus_surface);
+    if (last_requested_focus_surface.lock() != focus_surface)
+    {
+        last_requested_focus_surface = focus_surface;
+        auto new_active_surfaces = get_active_surfaces(focus_surface);
+
+        /// HACK: Grabbing popups (menus, in Mir terminology) should be given keyboard focus according to xdg-shell,
+        /// however, giving menus keyboard focus breaks Qt submenus. As of February 2022 Weston and other compositors
+        /// disobey the protocol by not giving keyboard focus to grabbing popups, so we do the same thing. The behavior
+        /// is subject change. See: https://github.com/MirServer/mir/issues/2324.
+        auto const new_keyboard_focus_surface = get_non_popup_parent(focus_surface);
+
+        notify_active_surfaces(lock, new_keyboard_focus_surface, new_active_surfaces);
+        set_keyboard_focus_surface(lock, new_keyboard_focus_surface);
+    }
+
     update_focus_locked(lock, focus_session, focus_surface);
 }
 
-void msh::AbstractShell::notify_focus_locked(
-    std::unique_lock<std::mutex> const& /*lock*/,
-    std::shared_ptr<ms::Surface> const& new_focus_surface)
+void msh::AbstractShell::notify_active_surfaces(
+    std::unique_lock<std::mutex> const&,
+    std::shared_ptr<ms::Surface> const& new_keyboard_focus_surface,
+    std::vector<std::shared_ptr<ms::Surface>> const& new_active_surfaces)
 {
-    auto const current_focus = notified_focus_surface.lock();
+    std::vector<std::shared_ptr<ms::Surface>> prev_active_surfaces;
+    prev_active_surfaces.reserve(notified_active_surfaces.size());
 
-    if (current_focus != new_focus_surface)
+    for (auto const& current_active_weak: notified_active_surfaces)
```

Comment:
I think this might be faster and clearer if we kept `notified_active_surfaces` and `new_active_surfaces` sorted, computed the set of surfaces that are in `notified_active_surfaces` but not `new_active_surfaces` in one pass rather than doing a bunch of O(n²) passes over them, and then doing the notifications at the end.